go/database/mpt/proof.go

```diff
@@ -115,3 +139,135 @@ func (p WitnessProof) AllStatesZero(root common.Hash, address common.Address, fr
 func (p WitnessProof) AllAddressesEmpty(root common.Hash, from, to common.Address) (tribool.Tribool, error) {
 	panic("not implemented")
 }
+
+// Equals returns true if the two witness proofs are equal.
+func (p WitnessProof) Equals(other WitnessProof) bool {
+	return maps.EqualFunc(p.proofDb, other.proofDb, rlpEncodedNodeEquals)
+}
+
+// MergeProofs merges the input witness proofs into the first witness proof.
+func MergeProofs(others ...WitnessProof) WitnessProof {
+	res := WitnessProof{make(proofDb)}
+	for _, other := range others {
+		res.Add(other)
+	}
+
+	return res
+}
+
+// visitWitnessPathTo visits all nodes from the input root following the input path.
+// Each encountered node is passed to the visitor.
+// If no more nodes are available on the path, the execution ends.
+// If the path does not exist, the function returns false.
+// The function returns an error if the path cannot be iterated due to error propagated from the input proof.
+// When the function reaches either an account node or a value node it is compared to the remaining input path
+// that was not iterated yet.
+// If the path matches, the function terminates and returns true.
+// It means this function can be used to find either an account node or a value node,
+// but it cannot find both at the same time.
+func visitWitnessPathTo(source proofDb, root common.Hash, path []Nibble, visitor witnessProofVisitor) (bool, error) {
+	nodeId := root
+
+	var nextEmbedded bool
+	var found, done bool
+	for !done {
+		var rlpNode rlpEncodedNode
+		if nextEmbedded {
+			rlpNode = nodeId[:]
+		} else {
+			var exists bool
+			rlpNode, exists = source[nodeId]
+			if !exists {
+				return false, nil
+			}
+		}
+		node, err := DecodeFromRlp(rlpNode)
+		if err != nil {
+			return false, err
+		}
+
+		var nextId common.Hash
+		switch n := node.(type) {
+		case *ExtensionNode:
+			if n.path.IsPrefixOf(path) {
+				nextId = n.nextHash
+				path = path[n.path.Length():]
+				nextEmbedded = n.nextIsEmbedded
+			} else {
+				done = true
+			}
+		case *BranchNode:
+			if len(path) == 0 {
+				done = true
+			} else {
+				nextId = n.hashes[path[0]]
+				nextEmbedded = n.isEmbedded(byte(path[0]))
+				path = path[1:]
+			}
+		case *AccountNode:
+			addressPath := createPathFromAddressPrefix(n.address, n.pathLength)
+			if addressPath.IsEqualTo(path) {
+				found = true
+			}
+			done = true
+		case *ValueNode:
+			keyPath := createPathFromKeyPrefix(n.key, n.pathLength)
+			if keyPath.IsEqualTo(path) {
+				found = true
+			}
+			done = true
+		default:
+			done = true
+		}
+
+		// visit when we are in the middle of the path or when we found the result
+		if !done || found {
+			visitor.Visit(nodeId, rlpNode, node, found && nextEmbedded)
+		}
```

Comment:
If I understand correctly, in all cases where the path was leading to a wrong node (e.g. an `ExtensionNode` with the wrong prefix or a leaf node with the wrong path), this final node is not included in the proof. Is this right?

This final "wrong direction" node, however, should be included. It is needed to proof that what was asked for is not present, and thus has the default value (zero, empty code, or empty storage).

Reply:
Yes, I thought that it was requirement not to include the final not matching node - I remember I was adding this condition after a review round. 

No problem to include the node, it actually simplifies the number of bool conditions. 

Reg. proving non-presence, i think it does not matter for proving a particular key, because when iterating the proof from root, the path will will not be found anyway does not matter if it is because the final node is not in the proof, or the node is not the fitting one.  

I guess a missing node may be needed for proving range of keys right?  We create a a proof for 
key1: 0xAA
key2: 0xAD

when we want to check this as a range, the now missing node could contain a path for 0xAB, 

right?

Reply:
ok, I think I updated the implementation to include the terminal nodes as well, except EmptyNode. 

It is also tested in the test: https://github.com/Fantom-foundation/Carmen/pull/904/files#diff-4324ecc3cdc67efb6e6fd6d00022d12c3345c7a1fdfe278090825c2a7062c60bR189